### PR TITLE
[SPARK-50033][SQL] Add a hint to logical.Aggregate() node

### PR DESCRIPTION
**Pull request by agubichev**
### What changes were proposed in this pull request?

Adds an abstract hint to the case class of Aggregate(). And adds an extension API to allow custom hint resolution rules.

### Why are the changes needed?

Allow users to define and resolve their own aggregate hints, and use the hints to guide the optimizing/planning behavior.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

a new test for custom aggregate hint

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by agubichev** [comment]
@cloud-fan

**Comment by HyukjinKwon** [comment]
Mind filing a JIRA please?

**Comment by agubichev** [comment]
> Mind filing a JIRA please?

Done in SPARK-50033

File: sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala [hunk 1]
```diff
@@ -543,6 +545,20 @@ class SparkSessionExtensionSuite extends SparkFunSuite with SQLHelper with Adapt
       }
     }
   }
+
+  test("test custom aggregate hint") {
+    // The custom hint allows us to replace the aggregate (without grouping keys) with just
+    // Literal.
+    withSession(Seq(_.injectHintResolutionRule(CustomerAggregateHintResolutionRule),
+      _.injectOptimizerRule(CustomAggregateRule))) { session =>
+      val res = ((session.range(10).agg(max("id")).as("max_id"))).
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SparkSessionExtensionSuite.scala:554]:
```suggestion
      val res = session.range(10).agg(max("id")).as("max_id").
```

**Comment by cloud-fan** [comment]
The pyspark failure is unrelated, thanks, merging to master!

**Pull request closed without merging**